src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -165,19 +171,40 @@ class Hyperspace(spark: SparkSession) {
   }
 }
 
-object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+object Hyperspace extends Logging {
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  // Value as (thread id, hyperspace context) to enforce one client thread per Spark session
+  // as Hyperspace doesn't guarantee thread safety.
+  private lazy val contexts =
+    new ConcurrentHashMap[SparkSession, (Long, HyperspaceContext)]().asScala
 
   private[hyperspace] def getContext(spark: SparkSession): HyperspaceContext = {
-    val ctx = context.get()
-    if (ctx == null || !ctx.spark.equals(spark)) {
-      // Ensure the Spark session the caller provides is the same as
-      // the one HyperspaceContext is using because Hyperspace depends on the
-      // session's properties such as configs, etc.
-      context.set(new HyperspaceContext(spark))
+    val ctx = contexts.get(spark)
+    val threadId = Thread.currentThread().getId
+    if (!ctx.isDefined) {
+      contexts.synchronized {
+        if (contexts.size >= HyperspaceConf.maxAllowedSparkSessions(spark)) {
+          throw HyperspaceException("Too many Spark sessions are using Hyperspace. " +
+            "Please increase the config `spark.hyperspace.maxNumAllowedConcurrentSparkSessions`. " +
+            s"Current limit: ${contexts.size}")
+        }
+        val newCtx = new HyperspaceContext(spark)
+        contexts.put(spark, (threadId, newCtx))
+        newCtx
+      }
+    } else {
+      if (ctx.get._1 != threadId) {
+        logWarning(s"Hyperspace is not thread safe for threads using one Spark session. " +
+          s"Please be aware of it. Current thread id: $threadId, Previous thread id: ${ctx.get._1}")
+        contexts.put(spark, (threadId, ctx.get._2))
+      }
+      ctx.get._2
     }
+  }
 
-    context.get()
+  def clearHyperspaceContexts(): Unit = {
```

Comment:
just `clear` seems fine? But, I am also fine if you think `clearHyperspaceContexts` is better.